### DISCUSSION TITLE
Trying to estimate drive storage for log retention.

### DISCUSSION THREAD

**Discussion by DebianGuru** [Security-Onion-Solutions/securityonion#6284] [Unsupported Versions] [answered]
First,  I apologize to have to do another thread about log retentions.  I have read many  threads about this and it seems to keep being a question that bugs a lot of people.

We were recently mandated by our new parent company to hold 6 months of logs for all traffic.  I've been running SO on a relatively small network as a standalone deployment for about a year.  There's 2 NICs capturing traffic and the management NIC.

I am doing full packet capture as well.  I understand that 6 months of full packet capture would be ridiculous, so I was hoping to get a month of full capture for my own use, but have 6 months worth of events logged to keep management happy.  I notice the 30,000 file limit cap for Stenographer, seems to be the limit on what I'm currently getting, because the 30k files takes me back to the 20+ days of pcaps I seem to be stuck at.  I seem to have a new file created each minute.  I should be able to crank that up to 46k or so to do about 32 days does this sound reasonable so far?  Am I on the right track?  (BTW, this is an EXT4 partition, so the 30k cap isn't a problem.)

I see about 45 previous days worth of Zeek logged connections in Kibana (There's always some kind of connections seen by Zeek, so I figured that was a good gauge to use to tell how far back my data goes.) 

I have 5.6TB of drive space in total and 2.5TB is currently used, leaving me with 2.8TB free.  1.4TB is currently in use by /nsm/elasticsearch. I also have about 780Gb of PCAPs.  That leaves me thinking that if I add the 2.8TB of free and add it to the 1.4TB by elasticsearch already, that should allow me 4,100GB (with conservative rounding) to enter in log_size_limit in the /opt/so/saltstack/local/pillar/minions/$minion.sls file.

I'm also confused by the terminology about hot indices in ES.  It says not to keep more than 30 days.  What is the difference between open/closed indices.  Sounds like the data is preserved on closed indices, but they can't be written to???

Am I on the right track?

Does that also tell me that it takes 1.4TB to store 45 days or 1.5 months (at least for my deployment).  I realize this will be based on traffic and can vary, but I need some way to estimate how much storage I'll need and then give a proof of concept to our executives.  If 1.4TB does 45 days, 6TB should cover 180days?  Should I estimate we need 8-10TB storage?  In my case I'll need another 1-3TB to my existing machine?

Thanks in advance.

**Comment by dougburks** [comment, marked as answer]
> I should be able to crank that up to 46k or so to do about 32 days does this sound reasonable so far?

Yes, you can increase as shown at https://docs.securityonion.net/en/2.3/stenographer.html#maximum-files.

> What is the difference between open/closed indices. Sounds like the data is preserved on closed indices, but they can't be written to???

Yes:
https://docs.securityonion.net/en/2.3/elasticsearch.html#closing-indices
https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-open-close.html

> Am I on the right track?

Seems like it.

**Reply by DebianGuru** [reply]
Doug,

Thanks for the reply. I'm not trying to beat a dead horse, just want to make sure I understand the way data is stored.  I don't want to tell our executives that I'm going to be able to store 6 months, then not be able to deliver on that.

https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-open-close.html says, "A closed index is blocked for read/write operations and does not allow all operations that opened indices allow. It is not possible to index documents or to search for documents in a closed index."

My global.sls file contains:
index_settings:
    so-redis:
      shards: 1
      warm: 7
      close: 30
      delete: 365

Does this mean that when I log into kibana, the data will not show up if I do a search from 5-6 months ago?  That's what it sounds like when I read, "A closed index is blocked for read/write operations."  Does this mean the search will just take longer to pull data than a search from a week ago?  My goal is just to ensure I can store 6 months of this information.

Will I need to change the line above from "close: 30" to "close: 180" in each statement of my global.sls file to ensure 6 months of data is retained?

Also, I was wanting someone to verify that my storage estimates sounded reasonable.

Thanks.

**Reply by dougburks** [reply]
> Does this mean that when I log into kibana, the data will not show up if I do a search from 5-6 months ago?

Yes, closed indices are not queried by Elasticsearch and so they will not return any data when searching.

> Will I need to change the line above from "close: 30" to "close: 180" in each statement of my global.sls file to ensure 6 months of data is retained?

Yes. However, please keep in mind the following from https://docs.securityonion.net/en/2.3/elasticsearch.html#closing-indices:

`The more indices are open, the more heap is required. Having too many open indices can lead to performance issues.`

**Reply by DebianGuru** [reply]
Doug,
Thanks for verifying these questions. This post contains my findings and the solution I used to restore the last 6 months of data for use in Kibana.  I have one question at the end if you don't care to read the whole post.

 I thought I was on the right track.  I did set the close time on those to 190 days.  I then wondered if I could test the performance without having to wait 5 months to see how things would work with 180 days worth of indices open. 

As a heads up, I'm running this on a substantial machine with two 12 core Xeons and 128GB of RAM.  I'm also on a small business sized network. I'm running this install in a standalone setup, because that's what I was allotted budget wise.  For anyone wanting to try this, see this also: [(https://docs.securityonion.net/en/16.04/elasticsearch.html#files)].  In summary, In my manager.sls file, under :elasticsearch, I set esheap: 24576m.  I set it to 24GB, though I have plenty of RAM, I there's a 32GB limit imposed by Java. 

I ran:

> curl -k https://localhost:9200/_cat/indices

to see all indices, then, piped that through grep like this:

> curl -k https://localhost:9200/_cat/indices |grep close

to see all the ones older than 30 days that curator had closed.
I then would run a nested for loop in the bash shell for each of the indices I wanted to open.
 

> for m in {01..12}; do for d in {01..31}; do curl -X  POST -k  "https://localhost:9200/so-zeek-2021.$m.$d/_open?pretty";done; done;

(Use whatever range for months and days you want to open.  I just used {05..31} for m and {01..31} for days.)

I did get an error at one point that said I had reached 1000 shards.  Apparently, there's a sane default of 1000 shards/node by default. 
I had to go in to Kibana, clicked the 3 lines in the upper left-hand corner and go to the management section and opened "dev tools".  In the left panel (where commands are entered, I put in:

> PUT /_cluster/settings
> {
>   "persistent" : {
>       "cluster.max_shards_per_node": "2000"
>   }
> }

Then put cursor back on the first line (PUT) then click the little triangular play icon to execute the command.

I was able to re-run my bash script to open all the indices that I wanted.  I overshot my 190 day range because I'd used 0..31 in my days variable and it was about the 20th of the month when I ran it.  I ran:

> curl -k https://localhost:9200/_cat/indices

after a few minutes and noticed that the files from 5/01 - 5/11 were closed again.... Curator had run and trimmed me back to 190 again... cool.

I repeated the script for other indices replacing "so-zeek" with the other indices like "so-syslog", "so-syslog", "so-strelka".......

So no Kibana goes back 190 days when I do any queries.  In my case, I don't see "performance issues".  Of course, when I do some queries, it does take longer (which is understandable), but it's acceptable to me.  I haven't seen anything fail to work yet.

Lastly, if you just need older logs, and don't need to have things searchable by Kibana (or maybe your setup isn't sufficient to handle that many indices open) you can see old logs in /nsm/zeek/logs. This contains logs that have been compiled and formatted in JSON format.  My install has logs from 9 months ago when I installed SecurityOnion.  Retention is likely dependent on traffic volume, and storage capacity.

So the only question I have is, "What setting will control when files in /nsm/zeek/logs will start being deleted?"  I'm assuming it will be tied to one of the settings that deals with free drive space.

Thanks again for all the help.  I hope this post helps someone else in my situation.

**Reply by dougburks** [reply]
> "What setting will control when files in /nsm/zeek/logs will start being deleted?" I'm assuming it will be tied to one of the settings that deals with free drive space.

/usr/sbin/so-sensor-clean runs via cronjob:
https://github.com/Security-Onion-Solutions/securityonion/blob/master/salt/common/tools/sbin/so-sensor-clean#L21
